File: src/main/java/com/github/pareronia/aoc/OCR.java
package com.github.pareronia.aoc;

import static com.github.pareronia.aoc.Utils.toAString;
import static java.util.Arrays.asList;

import java.util.HashMap;
import java.util.Map;
import java.util.NoSuchElementException;
import java.util.stream.Stream;

import com.github.pareronia.aoc.Grid.Cell;

public class OCR {

	private static final CharGrid A6 = CharGrid.from(asList(".##.",
													"#..#",
													"#..#",
													"####",
													"#..#",
													"#..#"));

	private static final CharGrid B6 = CharGrid.from(asList("###.",
													"#..#",
													"###.",
													"#..#",
													"#..#",
													"###."));

	private static final CharGrid C6 = CharGrid.from(asList(".##.",
													"#..#",
													"#...",
													"#...",
													"#..#",
													".##."));

	private static final CharGrid E6 = CharGrid.from(asList("####",
													"#...",
													"###.",
													"#...",
													"#...",
													"####"));

	private static final CharGrid F6 = CharGrid.from(asList("####",
													"#...",
													"###.",
													"#...",
													"#...",
													"#..."));

	private static final CharGrid G6 = CharGrid.from(asList(".##.",
													"#..#",
													"#...",
													"#.##",
													"#..#",
													".###"));
	
	private static final CharGrid H6 = CharGrid.from(asList("#..#",
													"#..#",
													"####",
													"#..#",
													"#..#",
													"#..#"));
	
	private static final CharGrid I6 = CharGrid.from(asList(".###",
													"..#.",
													"..#.",
													"..#.",
													"..#.",
													".###"));
	
	private static final CharGrid J6 = CharGrid.from(asList("..##",
													"...#",
													"...#",
													"...#",
													"#..#",
													".##."));
	
	private static final CharGrid K6 = CharGrid.from(asList("#..#",
													"#.#.",
													"##..",
													"#.#.",
													"#.#.",
													"#..#"));
	
	private static final CharGrid L6 = CharGrid.from(asList("#...",
													"#...",
													"#...",
													"#...",
													"#...",
													"####"));
	
	private static final CharGrid O6 = CharGrid.from(asList(".##.",
													"#..#",
													"#..#",
													"#..#",
													"#..#",
													".##."));
	
	private static final CharGrid P6 = CharGrid.from(asList("###.",
													"#..#",
													"#..#",
													"###.",
													"#...",
													"#..."));
	
	private static final CharGrid R6 = CharGrid.from(asList("###.",
													"#..#",
													"#..#",
													"###.",
													"#.#.",
													"#..#"));
	
	private static final CharGrid S6 = CharGrid.from(asList(".###",
													"#...",
													"#...",
													".##.",
													"...#",
													"###."));
	
	private static final CharGrid U6 = CharGrid.from(asList("#..#",
													"#..#",
													"#..#",
													"#..#",
													"#..#",
													".##."));
	
	private static final CharGrid Y6 = CharGrid.from(asList("#...",
													"#...",
													".#.#",
													"..#.",
													"..#.",
													"..#."));
	
	private static final CharGrid Z6 = CharGrid.from(asList("####",
													"...#",
													"..#.",
													".#..",
													"#...",
													"####"));
	
	@SuppressWarnings("serial")
	private static final Map<CharGrid, Character> glyphs = new HashMap<CharGrid, Character>() {
		{
			put(A6, 'A');
			put(B6, 'B');
			put(C6, 'C');
			put(E6, 'E');
			put(F6, 'F');
			put(G6, 'G');
			put(H6, 'H');
			put(I6, 'I');
			put(J6, 'J');
			put(K6, 'K');
			put(L6, 'L');
			put(O6, 'O');
			put(P6, 'P');
			put(R6, 'R');
			put(S6, 'S');
			put(U6, 'U');
			put(Y6, 'Y');
			put(Z6, 'Z');
		}

		@Override
		public Character get(Object grid) {
			if (!containsKey(grid)) {
				throw new NoSuchElementException(
						"OCR doesn't understand " + System.lineSeparator() + grid);
			}
			return super.get(grid);
		}
	};
	
	public static String convert6(CharGrid grid, Character fillChar, Character emptyChar) {
		return Stream.iterate(0,  i -> i + 5).limit(grid.getWidth() / 5)
				.map(i -> grid.subGrid(Cell.at(0, i), Cell.at(grid.getHeight(), i + 4)))
				.map(g -> g.replace(fillChar, '#'))
				.map(g -> g.replace(emptyChar, '.'))
				.map(glyphs::get)
				.collect(toAString());
	}
}

File: src/main/julia/ocr.jl
module OCR

A6 = Vector{String}([
    ".##.",
    "#..#",
    "#..#",
    "####",
    "#..#",
    "#..#"
   ])

B6 = Vector{String}([
    "###.",
    "#..#",
    "###.",
    "#..#",
    "#..#",
    "###."
   ])

C6 = Vector{String}([
    ".##.",
    "#..#",
    "#...",
    "#...",
    "#..#",
    ".##."
   ]);

E6 = Vector{String}([
    "####",
    "#...",
    "###.",
    "#...",
    "#...",
    "####"
   ])

F6 = Vector{String}([
    "####",
    "#...",
    "###.",
    "#...",
    "#...",
    "#..."
   ])

G6 = Vector{String}([
    ".##.",
    "#..#",
    "#...",
    "#.##",
    "#..#",
    ".###"
   ])

H6 = Vector{String}([
    "#..#",
    "#..#",
    "####",
    "#..#",
    "#..#",
    "#..#"
   ])

I6 = Vector{String}([
    ".###",
    "..#.",
    "..#.",
    "..#.",
    "..#.",
    ".###"
   ])

J6 = Vector{String}([
    "..##",
    "...#",
    "...#",
    "...#",
    "#..#",
    ".##."
   ])

K6 = Vector{String}([
    "#..#",
    "#.#.",
    "##..",
    "#.#.",
    "#.#.",
    "#..#"
   ])

L6 = Vector{String}([
    "#...",
    "#...",
    "#...",
    "#...",
    "#...",
    "####"
   ])

O6 = Vector{String}([
    ".##.",
    "#..#",
    "#..#",
    "#..#",
    "#..#",
    ".##."
   ])

P6 = Vector{String}([
    "###.",
    "#..#",
    "#..#",
    "###.",
    "#...",
    "#..."
   ])

R6 = Vector{String}([
    "###.",
    "#..#",
    "#..#",
    "###.",
    "#.#.",
    "#..#"
   ])

S6 = Vector{String}([
    ".###",
    "#...",
    "#...",
    ".##.",
    "...#",
    "###."
   ])

U6 = Vector{String}([
    "#..#",
    "#..#",
    "#..#",
    "#..#",
    "#..#",
    ".##."
   ])

Y6 = Vector{String}([
    "#...",
    "#...",
    ".#.#",
    "..#.",
    "..#.",
    "..#."
   ])

Z6 = Vector{String}([
    "####",
    "...#",
    "..#.",
    ".#..",
    "#...",
    "####"
   ])

GLYPHS = Dict{Vector{String},Char}(
    A6 => 'A',
    B6 => 'B',
    C6 => 'C',
    E6 => 'E',
    F6 => 'F',
    G6 => 'G',
    H6 => 'H',
    I6 => 'I',
    J6 => 'J',
    K6 => 'K',
    L6 => 'L',
    O6 => 'O',
    P6 => 'P',
    R6 => 'R',
    S6 => 'S',
    U6 => 'U',
    Y6 => 'Y',
    Z6 => 'Z',
)

function convert_6(grid::Vector{String}, fill::Char, empty::Char)
    glyphs = []
    for line ∈ grid
        push!(glyphs, replace(replace(line, fill => '#'), empty => '.'))
    end
    ans = ""
    for i ∈ 1:5:length(first(glyphs))
        glyph = []
        for line ∈ glyphs
            push!(glyph, line[i:i+3])
        end
        ans *= GLYPHS[glyph]
    end
    return ans
end

end  # module OCR
